sgkit/stats/pedigree.py

```diff
@@ -294,17 +290,113 @@ def kinship_diploid(
 
 
 @numba_jit
-def _inbreeding_as_self_kinship(
-    inbreeding: float, ploidy: int
-) -> float:  # pragma: no cover
+def _identify_founders_diploid(parent: ArrayLike) -> ArrayLike:
+    n = len(parent)
+    out = np.zeros(n, dtype=np.bool8)
+    for i in range(n):
+        if (parent[i, 0] < 0) or (parent[i, 1] < 0):
+            out[i] = True
+    return out
+
+
+@numba_jit
+def project_kinship_diploid(
+    parent: ArrayLike,
+    founder_kinship: ArrayLike,
+    founder_indices: ArrayLike,
+    allow_half_founders: bool = False,
+) -> ArrayLike:
+    """Project founder kinships along a pedigree to their decedents assuming
+    all individuals are diploid.
+
+    Parameters
+    ----------
+    parent
+        A matrix of shape (samples, parents) containing the indices of each
+        sample's parents with negative values indicating unknown parents as
+        defined in :data:`sgkit.variables.parent_spec`.
+    founder_kinship
+        A matrix of shape (founders, founders) containing kinship values among
+        founding samples.
+    founder_indices
+        An integer array of length founders mapping each founder to a sample
+        in the pedigree.
+    allow_half_founders
+        If False (the default) then a ValueError will be raised if any
+        individuals only have a single recorded parent.
+        If True then the unrecorded parent will be assumed to be
+        a unique founder unrelated to all other founders.
+
+    Note
+    ----
+    Any non-founders included in the founder_kinship and founder_indices arrays
+    will be treated as founders by ignoring their parentage.
+
+    Returns
+    -------
+    kinship
+        A square matrix of kinship estimates with self-kinship values on
+        the diagonal.
+
+    Raises
```

Comment:
I'm not sure how exact the docs need to be here, but this list is missing the AssertionErrors that can be raised by 354 and 355. (Same for the other function below)

Reply:
Good question, those assertions should never be raised via the public API (`pedigree_kinship`) because the dimensions are checked here: https://github.com/pystatgen/sgkit/blob/b966a1a76b486d5aa5ddc162d9b631d36c210e4e/sgkit/stats/pedigree.py#L1037. That check itself is also not strictly necessary as the dimensions are checked automatically against the `da.gufunc` signature here https://github.com/pystatgen/sgkit/blob/b966a1a76b486d5aa5ddc162d9b631d36c210e4e/sgkit/stats/pedigree.py#L1057 but this results in an error that may be unclear to the user (`ValueError: Dimension 'f' with different lengths in arrays`) hence the custom error message (or maybe I'm being overly cautious?). The assertions are really just there as a safety net for future development.

Reply:
Ah, nice! Thanks for the explanation. I think it is good to catch more obscure error messages as you've done here.